packages/core-data/src/selectors-ts.ts

```diff
@@ -0,0 +1,140 @@
+/**
+ * External dependencies
+ */
+import createSelector from 'rememo';
+import { set, get } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { getNormalizedCommaSeparable } from './utils';
+import type {
+	Context,
+	EntityQuery,
+	KeyOf,
+	Kind,
+	KindOf,
+	Name,
+	NameOf,
+	RecordOf,
+	DefaultContextOf,
+} from './types';
+
+// Placeholder State type for now
+export type State = any;
+
+/**
+ * Returns the Entity's record object by key. Returns `null` if the value is not
+ * yet received, undefined if the value entity is known to not exist, or the
+ * entity object if it exists and is received.
+ *
+ * @param  state State tree
+ * @param  kind  Entity kind.
+ * @param  name  Entity name.
+ * @param  key   Record's key
+ * @param  query Optional query.
+ *
+ * @return Record.
+ */
+export const getEntityRecord = createSelector(
+	function <
+		R extends RecordOf< K, N >,
+		C extends Context = DefaultContextOf< R >,
+		K extends Kind = KindOf< R >,
+		N extends Name = NameOf< R >,
+		Q extends EntityQuery< any > = EntityQuery< C >
+	>(
+		state: State,
+		kind: K,
+		name: N,
+		key: KeyOf< R >,
+		query?: Q
+	):
+		| ( Q[ '_fields' ] extends string[]
+				? Partial< RecordOf< K, N, C > >
```

Comment:
this is a great addition!

seeing `Partial` made me think we could go a step further and (for knowable cases) return an object that's not `Partial` but rather one that only has the requested fields.

surprisingly, I found a way to get TypeScript to infer that given the `_fields` value but I had the wrong type for `_fields` in my mind. I thought it was a comma-separated list of field names. I assumed it would be easier to do this given a list of field names, but I was wrong.

if people declare `_fields` with `as const` then we can do that. if not then I've only found a way to infer `string[]` as that field's type, meaning no information about the key names sadly.

Reply:
@dmsnell I played with that a bit more and got some success with the following types:

```ts
type ExtractKeys<
	Source,
	Options extends { _fields?: ( keyof Source )[] }
> = Options extends {
	_fields: infer KeysArray;
}
	? KeysArray extends Array< infer KeysUnion >
		? KeysUnion
		: never
	: never;

type PickMaybeFn = < Source, Options extends { _fields?: ( keyof Source )[] } >(
	obj: Source,
	options?: Options
) => ExtractKeys< Source, Options > extends never
	? Source
	: Pick< Source, ExtractKeys< Source, Options > >;

const pick: PickMaybeFn = ( obj, options? ) => { /* ... */ };

const obj = { a: 1, b: 2, c: 3 };

// Return the verbatim object without the Pick wrapper
const dontPick: typeof obj = pick( obj );
const dontPick2: typeof obj = pick( { a: 1, b: 2, c: 3 }, { _fields: [] } );

// Return the Picked version of an object
const pickA: Pick< typeof obj, 'a' > = pick(
	{ a: 1, b: 2, c: 3 },
	{ _fields: [ 'a' ] }
);
const pickAB: Pick< typeof obj, 'a' | 'b' > = pick(
	{ a: 1, b: 2, c: 3 },
	{ _fields: [ 'a', 'b' ] }
);

// TS Error: 'c' is missing in type
const doNotPickAbc: Pick< typeof obj, 'a' | 'b' | 'c' > = pick(
	{ a: 1, b: 2, c: 3 },
	{ _fields: [ 'a', 'b' ] }
);
```

[Playground link](https://www.typescriptlang.org/play?ssl=28&ssc=1&pln=29&pc=1#code/C4TwDgpgBAogHsATgQwMbANIRAZwDwBQAkAMoD2AroqhADTEDyYwAlmQHY5QQITsAmXAN5QA+gDMWEADaCA-AC4oACigBrbGXFRyVGlACUAbQC6UAL4EAfFAC8UJqw5cewPoKhDiEqbJxKWdnEIRCgsXABBRBQQAG4CSyI5MOwcKJjuXgEudOQQPChA4NDwnABVdjZ2KCtiJJTcCqq6pXYIADcQ4laOkPiCUEgoAAUWVDUAWTyAIwgAMWr7At1qOgdmKpcsjxEfGXklVQ0QLR1KVcNTCxrlYjJpgCslFZp6IjIN50V1p04CAzsNngSDQmFSy3Orx+mxqmTc2SgbU6iGIyReEG6IzGaghejWwJQ6FKuNWtGhzlhVn6qGcwCgYGxSlG4ymIFmCzsKig9weZI+vxwyQBthsXiILG0qgAhNzPpw5AA6PZ+RXSPgAc2AAAtDJ46ogIMAqNUefEiIkaZw6QzxhB+EpkOwQJyhOYzeIyKFVJacHTjtztPzNkrJPsuACxUQbRp+EZ1NgoGZ7Dy4-6TGbEgajYhqtG7fE3QQCD66TyXVBkEoAIxk6ZKABMZNQSgAzBZ+gB6DtQABKhuNUG10GR02QrAAttzHhB0FAAO4sbWUOlDrHjecoMCQFElqD8DjAZlqJSDCCnMv2aOqMsGeK7-fsQ-Y+sn8BnwOPTlXzwV6u1htNq2FhkrsoZ+EoVzmIYnbdn22bVKuR52lAyI4FUAYVia07oMWtL0tiERMtiBSnuejxkgA5MgFGwpe2K3EQIiVlANZQHWUCNlAzZQG25hvKBviCBBUBUTRZiWLeuFWvh4wRAAQkR4wkW+ZG8iJ1FQAAPiJ0w0TYdHjAxTF-mxAFcUBfHEAJYbCaJlG6YmFj-DBUAACokLA0SekoFGoDRLBcOOAVoew6qFAhb5Sb6e5kAAcmQT6ydM3FHspkCqZRGnaRRDnZX5tEyWoRm-ix-4cYBPHAVZYhgUJUBxnZOliU5kkEF2UAAJKcJAs6rgaOAUNIwCBY6FDINI0ggAoUV0oEOA9cAEwQFWX70T+zGsexnHcbx0EzeF80zotED1qthnrSZW0VbxIE1YJ-j1eJe27nNC1LW2BlFRdpWmeV5mVXxP7KnVDXUfZzVQa1PpkGqCrSGQ6rKEID6JWoZIo0enHRhEZLY3J5gGEAA)

I

Reply:
@dmsnell I added support for `Pick` to this PR. Unfortunately I couldn't figure out a way to retain the `Context` as the second generic parameter. This is the simplest signature I came up with:

```ts
	function <
		R extends RecordOf< K, N >,
		Query extends EntityQuery< K, N, InferredContext >,
		K extends Kind = KindOf< R >,
		N extends Name = NameOf< R >,
		InferredContext extends Context = Query extends { context: infer C }
			? C
			: DefaultContextOf< R >
	>(
		state: State,
		kind: K,
		name: N,
		key: KeyOf< R >,
		query?: Query
	): PickQueryFields< RecordOf< K, N, InferredContext >, Query > | undefined {
```

Since TS type inference is all or nothing, passing a type for R requires the query type, too:

```ts
const comment = getEntityRecord<
	Comment< 'view' >,
	{ context: 'view'; _fields: [ 'id' ] }
>( {}, 'root', 'comment', 15, { context: 'view', _fields: [ 'id' ] } );
// comment is Pick< Comment< 'view' >, 'id' >
```

Perhaps this isn't a big issue, though. I can't think of a case where you wouldn't want to rely on inference given that `getEntityRecord` only supports the registered entity types with known kind and name.

Reply:
leaving it at `Partial` for now is probably a good place to be.

in your playground link you can see the same trouble I faced when not using type-narrowing literals. Hint: if you pass your objects as literals in the function invocation or type parameter it will type differently than if you pass in an object defined elsewhere.

```ts
// this works
// Return the Picked version of an object
pick(t, { _fields: ['a', 'b'] });
```

```ts
// this doesn't work
const q = { _fields: [ 'a', 'b' ] };

// Return the Picked version of an object
pick(t, q);
```

Because it can't assert that `q._fields` are keys of the object. We can add some `readonly` attributes to the `(keyof Source)[]` types and that actually let's us do this, but there's a caveat. We have to define `q` with `as const`, or at lest the `_fields` property `as const`.

https://user-images.githubusercontent.com/5431237/156456188-6f24a6c6-3b86-4153-9cfc-54c93f43bf07.mov

There may still be a way to get this, but I consider it a non-starter if we have to have people declare those fields `as const` manually and in every invocation.

Reply:
The core problem I think is getting a list of strings into a type union of those string values without needing `as const`. This illustrates the problem as `Fields` and `Fields2` both resolve to `string` instead of as `'a' | 'b' | 'id' | 'name'`

```ts
const l = ['a', 'b', 'id', 'name'];

type Names<T extends any[]> = T[number];
type Names2<T extends any[]> = T extends (infer V)[] ? V : never;

type Fields = Names<typeof l>;
type Fields2 = Names2<typeof l>;
```

Reply:
Ah great point @dmsnell! It seems like all the type information is gone right after the list declaration in the current TypeScript version, so this isn't something we can fix. The best we can do here is return `Pick<>` when the keys are known, and `Partial<>` when they aren't. For now, let's abandon `Pick` and go with Partial as you suggested – `Pick` is a new feature that can be added later on.